---
title: "Example `ChIPpeakAnno`"
author: "Maurits Evers"
output:
  html_document:
#    code_folding: hide
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
    theme: united
    df_print: paged
    fig_width: 8
    fig_height: 6
---


```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```


# Load libraries

```{r load-libraries}
library(ChIPpeakAnno);
library(GenomicRanges);
library(TxDb.Hsapiens.mevers.hs1);
library(tidyverse);
```


# Read ENCODE KDM4A peak data

```{r read-peak-data}
# Read ENCODE KDM4A sample data
# https://www.encodeproject.org/experiments/ENCSR000AVC/
peaks <- read.table("ENCFF205WRX.bed", header = FALSE);
peaks <- with(peaks, GRanges(seqnames = V1, IRanges(start = V2, end = V3)));
```

Link to ENCODE experiment summary:  [https://www.encodeproject.org/experiments/ENCSR000AVC](https://www.encodeproject.org/experiments/ENCSR000AVC/)

# Extract gene annotation from `TxDb` object

```{r extract-gene_annotation}
txdb <- TxDb.Hsapiens.mevers.hs1;
annoData <- toGRanges(txdb, feature = "gene");
```


# Meta-gene plot

```{r calculate-meta-gene, fig.show='hide'}
# Meta-gene plot by aggregating peaks across genes
df <- binOverFeature(
    peaks, 
    annotationData = annoData,
    radius = 5000,
    nbins = 100,
    FUN = length,
    errFun = 0,
    ylab="count",
    main="Distribution of aggregated peak numbers around TSS");
```

```{r plot-meta-gene}
# Plot
df %>%
    as.data.frame() %>%
    rownames_to_column("distance") %>%
    mutate(distance = as.numeric(as.character(distance))) %>%
    ggplot(aes(distance, peaks)) +
    geom_line() +
    labs(x = "Distance from TSS", y = "Count") +
    theme_bw();

```

# Feature-specific binding site abundances

```{r plot-peak-abundance}
# Match seqlevels 
# If peaks contains features in seqlevels that are not part of txdb,
# assignChromosomeRegion will throw an error.
common.seqlevels <- intersect(seqlevels(txdb), seqlevels(peaks));
peaks <- keepSeqlevels(peaks, common.seqlevels, pruning.mode = "coarse");

# Summarise peak abundances per genomic feature (exon, intron, etc.)
res <- assignChromosomeRegion(
    peaks,
    nucleotideLevel = FALSE,
    precedence = c("Promoters", "fiveUTRs", "threeUTRs", "Exons", "Introns"),
    TxDb = TxDb.Hsapiens.mevers.hs1)

# Plot
res$percentage %>%
    as.data.frame() %>%
    ggplot(aes(subjectHits, Freq)) + 
    geom_bar(stat = "identity") + 
    labs(x = "Feature", y = "Percentage") + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
```